### PR TITLE
🚀 Docker image for easier deployment

### PR DESCRIPTION
**Pull request by aledjones**
- [x] Dockerfile: creates image with prebuilt webpack
- [x] Dockerfile: has maintainer labels
- [x] Dockerfile: use composer image
- [x] docker-entrypoint.sh: migrates current database on start
- [ ] docker-entrypoint.sh: locking mechanism for database migration
- [ ] docker-entrypoint.sh: error-handling
- [ ] docker-entrypoint.sh: locking state validation
- [x] GitHub Actions: Images are built on push to defined branches
- [x] GitHub Actions: Images are built and tagged with commit id
- [x] GitHub Actions: Images are pushed to GitHub Registry tagged with the gut reference that started the build (branch push or tag release)

### CODE REVIEW

File: Dockerfile [hunk 1]
```diff
@@ -0,0 +1,27 @@
+FROM node:16 as NodeBuildContainer
+COPY . /usr/src/trwl
+WORKDIR /usr/src/trwl
+RUN npm i
+RUN npm run prod
+
+FROM php as PHPBuildContainer
```
Comment (MrKrisKrisu) [inline Dockerfile]: Can you specify PHP 8.0?

File: app/Providers/AppServiceProvider.php [hunk 2]
```diff
@@ -26,6 +27,9 @@ public function register(): void {
      * @throws BindingResolutionException
      */
     public function boot(): void {
+        if (env('APP_ENV') === 'prod') {
```
Comment (MrKrisKrisu) [inline app/Providers/AppServiceProvider.php]:
```suggestion
        if (config('app.env') === 'prod') {
```

File: docker-entrypoint.sh [hunk 3]
```diff
@@ -0,0 +1,12 @@
+#!/bin/bash
+set -e
+
+if [ "$1" = 'apache2-foreground' ]; then
+    wait-for-it $DB_HOST:${DB_PORT:=3306}
+    cd /var/www/html
+    runuser -u www-data -- php artisan migrate
```
Comment (MrKrisKrisu) [inline docker-entrypoint.sh]:
```suggestion
    runuser -u www-data -- php artisan migrate --force
```

Isn't this required?

File: Dockerfile [hunk 4]
```diff
@@ -0,0 +1,27 @@
...
+RUN php -r "copy('https://getcomposer.org/installer', 'composer-setup.php');" && \
+    php composer-setup.php --install-dir=/bin --filename=composer
+RUN apt update && apt install -y zlib1g-dev libpng-dev zip
+RUN docker-php-ext-install gd exif
```
Comment (MrKrisKrisu) [inline Dockerfile]: We have already talked about that. In the long run, we should use a composer image and set a version. I have written it on the todo. But it is non-blocking for me.

File: Dockerfile [hunk 5]
```diff
@@ -0,0 +1,27 @@
...
+
+COPY --from=NodeBuildContainer /usr/src/trwl /usr/src/trwl
+WORKDIR /usr/src/trwl
+RUN composer install
+
+FROM php:apache
```
Comment (MrKrisKrisu) [inline Dockerfile]: Can you specify PHP 8.0?

File: .htaccess [hunk 6]
```diff
@@ -0,0 +1,6 @@
+Options +FollowSymLinks
+RewriteEngine On
+
+RewriteCond %{REQUEST_FILENAME} !-d
+RewriteCond %{REQUEST_FILENAME} !-f
+RewriteRule ^ index.php [L]
```
Comment (MrKrisKrisu) [inline .htaccess]: What is the file needed for? It is located in the root directory.

In the public folder I would understand it.

File: .github/workflows/docker.yml [hunk 7]
```diff
@@ -0,0 +1,48 @@
+# shamelessly copied from https://docs.github.com/en/packages/managing-github-packages-using-github-actions-workflows/publishing-and-installing-a-package-with-github-actions#upgrading-a-workflow-that-accesses-ghcrio
+# modified to meet our needs
+name: "Docker Image"
+
+on:
+  push:
+    branches:
+      - dev-*
```
Comment (MrKrisKrisu) [inline .github/workflows/docker.yml]: I would not build a container on our dev-* branches. Or do you think that makes sense?

File: .dockerignore [hunk 8]
```diff
@@ -0,0 +1,3 @@
+.git
```
Comment (MrKrisKrisu) [inline .dockerignore:1]: maybe also add vendor/ and node_modules? This would at least be useful for local building.

**Review by MrKrisKrisu — CHANGES_REQUESTED** [review]

File: .dockerignore (hunk 8 above)
Comment (aledjones) [reply .dockerignore:1]: Will do

File: .github/workflows/docker.yml (hunk 7 above)
Comment (aledjones) [reply .github/workflows/docker.yml]: This was only a suggestion for easier local and on-premise testing like in cloud environments...
I'll remove those entries as they can be added back in the future

Comment (aledjones) [reply .github/workflows/docker.yml]: On second thought: this is the only way to test docker builds in my PR branch.
It'll stay and be removed as soon as we merge into develop.

File: .htaccess (hunk 6 above)
Comment (aledjones) [reply .htaccess]: This file just shows that I'm an idiot that didn't put together a laravel container for a very long time.
Will be removed in upcoming changes.

File: Dockerfile (hunk 1 above)
Comment (aledjones) [reply Dockerfile]: Will do.

File: Dockerfile (hunk 4 above)
Comment (aledjones) [reply Dockerfile]: Changes to `composer:2` in upcoming commit

File: app/Providers/AppServiceProvider.php [hunk 9]
```diff
@@ -26,6 +27,9 @@ public function register(): void {
      * @throws BindingResolutionException
      */
     public function boot(): void {
+        if (config('app.env') === 'prod') {
```
Comment (MrKrisKrisu) [inline app/Providers/AppServiceProvider.php]:
```suggestion
        if (config('app.env') === 'production') {
```
Sorry, I missed that last time. xD

**Review by MrKrisKrisu — CHANGES_REQUESTED** [review]

**Review by MrKrisKrisu — APPROVED** [review]

**Pull request merged**